File: shpritz-node/shpritz.js
/*jslint esnext: true:*/

let minDist = 30;
let trig = false;

let app = require('express')(); 
let http = require('http').Server(app);
let io = require('socket.io')(http);

app.get('/', function(req, res){
	res.sendFile(__dirname + '/dashboard.html');
});

http.listen(5000, function(){
	console.log('listening on *:5000');
});

let SerialPort = require('serialport');
let arduino = new SerialPort('/dev/ttyACM0', {
	baudRate: 9600
});



arduino.open((err) => {
	return console.log(err.message);
});

arduino.on('data', (data) => {
	// data value is distance in cm
	if (!trig) {
		let dist = data[data.length - 1];
		io.emit('dashboard', {'distance': dist});
		setShpritz(dist < minDist);
	}
});

function setShpritz(state) {
	arduino.write(state ? '1' : '0');
	io.emit('dashboard', {'shpritz': state});
}


io.on('connection', function(socket) {
	socket.on('dashboard', (msg) => {
		if (msg.hasOwnProperty('man_trigger')) {
			trig = msg.man_trigger;
			setShpritz(trig);
		}
	});	
});

File: shpritz-python/shpritz.py
import socket
import serial
import json
import threading

thresh = 30
trig = False

arduino = serial.Serial('/dev/ttyACM0', 9600)  # open serial port


udp = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
udp.bind(("127.0.0.1", 3001))

def udp_thread():
  global trig
  while True:
    data, addr = udp.recvfrom(1024) # buffer size is 1024 bytes
    msg = json.loads(data)
    if "man_trigger" in msg:
      trig = msg["man_trigger"]
      setShpritz(trig)
    
def setShpritz(state):
  if state:
    arduino.write('1')
  else:
    arduino.write('0')
  message = json.dumps({"shpritz": state})
  udp.sendto(message, ("127.0.0.1", 3002))
  
t = threading.Thread(target=udp_thread)
t.start()

if __name__ == '__main__':
  while True:
    distance = ord(arduino.read())
    if not trig:
      setShpritz(distance < thresh)
    message = {"distance": distance}
    udp.sendto(json.dumps(message), ("127.0.0.1", 3002))
    
